### PR TITLE
Add order and subok argument to array function

### PR DESCRIPTION
**Pull request by okuta**
Rel #158 .

### CODE REVIEW

**Comment by niboshi** [comment]
Please fix conflict.
LGTM otherwise.

**Comment by okuta** [comment]
@niboshi I resolved conflicts.

**Comment by niboshi** [comment]
Oh, sorry for delay.
Jenkins, test this please

**Comment by niboshi** [comment]
@okuta Please check test results.

**Comment by niboshi** [comment]
I think these test cases cannot check if the order is correct.
How about adding an option to `@testing.numpy_cupy_array_equal()` to check the order as well?

**Comment by okuta** [comment]
I will try it.

**Comment by niboshi** [comment]
FYI: Jenkins test failed. Is it relevant to this PR?

**Comment by niboshi** [comment]
jenkins, test this please

File: cupy/testing/array.py [hunk 1]
```diff
@@ -89,6 +89,18 @@ def assert_array_equal(x, y, err_msg='', verbose=True):
...
         cupy.asnumpy(x), cupy.asnumpy(y), err_msg=err_msg,
         verbose=verbose)
+    if (isinstance(x, (numpy.ndarray, cupy.ndarray)) and
+            isinstance(y, (numpy.ndarray, cupy.ndarray))):
+        if x.flags.c_contiguous != x.flags.c_contiguous:
+            raise AssertionError(
+                'The state of c_contiguous flag is different. '
+                '(x:{} y:{})'.format(x.flags.c_contiguous,
+                                     y.flags.c_contiguous))
+        if x.flags.c_contiguous != x.flags.c_contiguous:
```
Comment (niboshi) [inline cupy/testing/array.py]: Should it be `if x.flags.f_contiguous != x.flags.f_contiguous`?

**Comment by okuta** [comment]
I fixed!

**Comment by okuta** [comment]
@niboshi Cloud you check this PR?

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -2011,19 +2014,22 @@ cdef _argmax = create_reduction_func(
...
         src = obj
         if dtype is None:
             dtype = src.dtype
         dev = src.data.device
         if dev is None or dev.id == device.get_device_id():
-            a = src.astype(dtype, copy=copy)
+            a = src.astype(dtype, order=order, copy=copy)
         else:
-            a = src.copy().astype(dtype, copy=False)
+            a = src.copy(order=order).astype(dtype, copy=False)
```
Comment (niboshi) [inline cupy/core/core.pyx]: Why can't we write `src.astype(dtype, order=order, copy=True)`?

File: cupy/testing/helper.py [hunk 3]
```diff
@@ -150,6 +150,8 @@ def numpy_cupy_allclose(rtol=1e-7, atol=0, err_msg='', verbose=True,
          sp_name(str or None): Argument name whose value is either
              ``scipy.sparse`` or ``cupy.sparse`` module. If ``None``, no
              argument is given for the modules.
+         contiguous_check(book): If ``True``, consistency of contiguous is
```
Comment (niboshi) [inline cupy/testing/helper.py]: typo: book -> bool

File: cupy/testing/helper.py [hunk 4]
```diff
@@ -173,8 +175,21 @@ def numpy_cupy_allclose(rtol=1e-7, atol=0, err_msg='', verbose=True,
     .. seealso:: :func:`cupy.testing.assert_allclose`
     """
     def check_func(cupy_result, numpy_result):
-        array.assert_allclose(cupy_result, numpy_result,
-                              rtol, atol, err_msg, verbose)
+        c = cupy_result
+        n = numpy_result
+        array.assert_allclose(c, n, rtol, atol, err_msg, verbose)
+        if (contiguous_check and
+                isinstance(c, cupy.ndarray) and isinstance(n, numpy.ndarray)):
```
Comment (niboshi) [inline cupy/testing/helper.py]: Why do you have to check types here?
If we need to check it, shouldn't it be checked regardless of `contiguous_check`?

File: cupy/testing/helper.py [hunk 5]
```diff
@@ -129,7 +129,7 @@ def test_func(self, *args, **kw):
 
 def numpy_cupy_allclose(rtol=1e-7, atol=0, err_msg='', verbose=True,
                         name='xp', type_check=True, accept_error=False,
-                        sp_name=None):
+                        sp_name=None, contiguous_check=True):
```
Comment (niboshi) [inline cupy/testing/helper.py]: How about `contiguous_check=False` by default? `np.allclose` does not check it.

File: cupy/manipulation/rearrange.py [hunk 6]
```diff
@@ -47,7 +47,7 @@ def fliplr(a):
     """
     if a.ndim < 2:
         raise ValueError('Input must be >= 2-d')
-    return cupy.take(a, cupy.arange(a.shape[1] - 1, -1, -1), axis=1)
```
Comment (niboshi) [inline cupy/manipulation/rearrange.py]: Is this related to this PR?

File: cupy/manipulation/rearrange.py [hunk 7]
```diff
@@ -67,7 +67,7 @@ def flipud(a):
     """
     if a.ndim < 1:
         raise ValueError('Input must be >= 1-d')
-    return cupy.take(a, cupy.arange(a.shape[0] - 1, -1, -1), axis=0)
```
Comment (niboshi) [inline cupy/manipulation/rearrange.py]: Ditto.

File: cupy/core/core.pyx (hunk 2 above)
Comment (okuta) [reply cupy/core/core.pyx]: Because this case is device to device copy.

File: cupy/testing/helper.py (hunk 4 above)
Comment (okuta) [reply cupy/testing/helper.py]: This branch supports single value or list.

File: cupy/core/core.pyx (hunk 2 above)
Comment (niboshi) [reply cupy/core/core.pyx]: I think `astype` copies to the current device if `copy` is `True`. Is that not correct?

File: cupy/testing/helper.py (hunk 4 above)
Comment (niboshi) [reply cupy/testing/helper.py]: Could you elaborate?
Do you mean `c` and `n` could be lists?

File: cupy/core/core.pyx (hunk 2 above)
Comment (okuta) [reply cupy/core/core.pyx]: `astype` only uses  `elementwise_copy`. It is not support multi GPUs copy.

File: cupy/testing/helper.py (hunk 4 above)
Comment (okuta) [reply cupy/testing/helper.py]: Yes. When we use `parameterize` test, sometimes NumPy returns not ndarray type.

File: cupy/core/core.pyx (hunk 2 above)
Comment (niboshi) [reply cupy/core/core.pyx]: OK.

File: cupy/testing/helper.py [hunk 8]
```diff
@@ -152,6 +152,8 @@ def numpy_cupy_allclose(rtol=1e-7, atol=0, err_msg='', verbose=True,
          sp_name(str or None): Argument name whose value is either
              ``scipy.sparse`` or ``cupy.sparse`` module. If ``None``, no
              argument is given for the modules.
+         contiguous_check(boool): If ``True``, consistency of contiguousness is
```
Comment (niboshi) [inline cupy/testing/helper.py]: Typo: boool

File: cupy/creation/from_data.py [hunk 9]
```diff
@@ -13,17 +13,33 @@ def array(obj, dtype=None, copy=True, ndmin=0):
         dtype: Data type specifier.
         copy (bool): If ``False``, this function returns ``obj`` if possible.
             Otherwise this function always returns a new array.
+        order ({'C', 'F', 'A', 'K'}): Row-major (C-style) or column-major
+            (Fortran-style) order.
+            When ``order`` is 'A', it uses 'F' if ``a`` is column-major and
+            uses 'C' otherwise.
+            And when ``order`` is 'K', it keeps strides as closely as
+            possible.
+            If ``obj`` is ``numpy.ndarray``, thi function returns 'C' or 'F'
```
Comment (niboshi) [inline cupy/creation/from_data.py]: * ``` ``numpy.ndarray`` ``` -> ``` :class:`numpy.ndaaray` ```
* typo: thi

File: cupy/creation/from_data.py [hunk 10]
```diff
@@ -13,17 +13,33 @@ def array(obj, dtype=None, copy=True, ndmin=0):
...
             head of the shape if needed.
 
     Returns:
         cupy.ndarray: An array on the current device.
 
+
+
+    .. note::
+       This method currently does not support ``subok``
+       arguments.
```
Comment (niboshi) [inline cupy/creation/from_data.py]: typo: arguments -> argument

**Comment by niboshi** [comment]
Please fix flake8 error.
LGTM except that.

**Comment by okuta** [comment]
Thank you. I fixed.

**Comment by niboshi** [comment]
LGTM!
